pilot/pkg/serviceregistry/kube/controller/controller.go

```diff
@@ -419,6 +421,42 @@ func (c *Controller) deleteService(svc *model.Service) {
 	c.handlers.NotifyServiceHandlers(nil, svc, event)
 }
 
+// recomputeServiceForPod is called when a pod changes and service endpoints need to be recompute
+// Most of Pod is immutable, so once it has been created we are ok to cache the internal representation.
+// However, a few fields (labels) are mutable. When these change, we call recomputeServiceForPod and rebuild the cache
+// for all service's the pod is a part of and push an update.
+func (c *Controller) recomputeServiceForPod(pod *v1.Pod) {
+	allServices := c.services.List(pod.Namespace, klabels.Everything())
+	cu := sets.New[model.ConfigKey]()
+	services := getPodServices(allServices, pod)
+	for _, svc := range services {
+		hostname := kube.ServiceHostname(svc.Name, svc.Namespace, c.opts.DomainSuffix)
+		c.Lock()
+		conv, f := c.servicesMap[hostname]
+		c.Unlock()
+		if !f {
+			return
+		}
+		shard := model.ShardKeyFromRegistry(c)
+		endpoints := c.buildEndpointsForService(conv, true)
+		if len(endpoints) > 0 {
```

Comment:
When a pod changes labels we call this function and find each service, then rebuild the endpoints. So this is triggered on pod label change for each service with >0 endpoints

Reply:
Sorry, I meant to ask when would it be zero and why do we need this condition? If it is zero what happens?

Reply:
Ah. If its 0 it should have been 0 before so this should be a NO-OP and minor optimization. I copy+pasted it from https://github.com/istio/istio/blob/ff923fcc18b68ed1d1329970f49b9ef370aa0a57/pilot/pkg/serviceregistry/kube/controller/controller.go#L500.

TBH I am not sure L500 is correct, seems like it could miss us intentionally scaling to 0. But maybe some other update would trigger it anyways

Reply:
>TBH I am not sure L500 is correct, seems like it could miss us intentionally scaling to 0. But maybe some other update would trigger it anyways

Yeah. That is what I think too.